include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,546 @@
+    return async_generator_increment_operation<T>{*this};
+  }
+
+  reference operator*() const noexcept { return m_coroutine.promise().value(); }
+
+  bool operator==(const async_generator_iterator& other) const noexcept {
+    return m_coroutine == other.m_coroutine;
+  }
+
+  bool operator!=(const async_generator_iterator& other) const noexcept {
+    return !(*this == other);
+  }
+
+private:
+  friend class async_generator_increment_operation<T>;
+
+  handle_type m_coroutine;
+};
+
+template <typename T>
+async_generator_iterator<T>&
+async_generator_increment_operation<T>::await_resume() {
+  if (m_promise->finished()) {
+    // Update iterator to end()
+    m_iterator = async_generator_iterator<T>{nullptr};
+    m_promise->rethrow_if_unhandled_exception();
+  }
+
+  return m_iterator;
+}
+
+template <typename T>
+class async_generator_begin_operation final
+  : public async_generator_advance_operation {
+  using promise_type = async_generator_promise<T>;
+  using handle_type = unifex::coro::coroutine_handle<promise_type>;
+
+public:
+  async_generator_begin_operation(std::nullptr_t) noexcept
+    : async_generator_advance_operation(nullptr) {}
+
+  async_generator_begin_operation(handle_type producerCoroutine) noexcept
+    : async_generator_advance_operation(
+          producerCoroutine.promise(), producerCoroutine) {}
+
+  bool await_ready() const noexcept {
+    return m_promise == nullptr ||
+        async_generator_advance_operation::await_ready();
+  }
+
+  async_generator_iterator<T> await_resume() {
+    if (m_promise == nullptr) {
+      // Called begin() on the empty generator.
+      return async_generator_iterator<T>{nullptr};
+    } else if (m_promise->finished()) {
+      // Completed without yielding any values.
+      m_promise->rethrow_if_unhandled_exception();
+      return async_generator_iterator<T>{nullptr};
+    }
+
+    return async_generator_iterator<T>{
+        handle_type::from_promise(*static_cast<promise_type*>(m_promise))};
+  }
+};
+}  // namespace detail
+
+template <typename T>
+class [[nodiscard]] async_generator {
+public:
+  using promise_type = detail::async_generator_promise<T>;
+  using iterator = detail::async_generator_iterator<T>;
+
+  async_generator() noexcept : m_coroutine(nullptr) {}
+
+  explicit async_generator(promise_type& promise) noexcept
+    : m_coroutine(
+          unifex::coro::coroutine_handle<promise_type>::from_promise(promise)) {
+  }
+
+  async_generator(async_generator&& other) noexcept
+    : m_coroutine(other.m_coroutine) {
+    other.m_coroutine = nullptr;
+  }
+
+  ~async_generator() {
+    if (m_coroutine) {
+      m_coroutine.destroy();
+    }
+  }
+
+  async_generator& operator=(async_generator&& other) noexcept {
+    async_generator temp(std::move(other));
+    swap(temp);
+    return *this;
+  }
+
+  async_generator(const async_generator&) = delete;
+  async_generator& operator=(const async_generator&) = delete;
+
+  auto next() noexcept {
```

Comment:
`next()` and `cleanup()` is as shared by @ispeters in https://godbolt.org/z/Kh3rWTMao